### PR TITLE
net/dns: make exit node DNS ask OSConfigurator for backup resolvers

### PR DESCRIPTION
**Pull request by bradfitz**
Updates #1713

/cc @crawshaw

### CODE REVIEW

File: net/dns/resolved.go [hunk 1]
```diff
@@ -332,6 +333,10 @@ func (m *resolvedManager) GetBaseConfig() (OSConfig, error) {
 	return OSConfig{}, ErrGetBaseConfigNotSupported
 }
 
+func (m *resolvedManager) GetExitNodeForwardResolver() ([]dnstype.Resolver, error) {
+	return []dnstype.Resolver{{Addr: "127.0.0.53:53"}}, nil
```
Comment (crawshaw) [inline net/dns/resolved.go:337]: Here's hoping no-one ever sets `DNSStubListener=no` in resolved.conf.

**Review by crawshaw — APPROVED** [review]

File: net/dns/resolved.go (hunk 1 above)
Comment (bradfitz) [reply net/dns/resolved.go:337]: That case will be fixed in a follow-up CL that also makes exit node DNS proxy work on Windows: using gethostbyname etc (via Go net package DNS lookups).

Comment (danderson) [reply net/dns/resolved.go:337]: FWIW: nobody runs resolved and disables the stub listener. It'd break too much stuff.

File: net/dns/manager.go [hunk 2]
```diff
@@ -62,6 +62,11 @@ func (m *Manager) Set(cfg Config) error {
 	if err != nil {
 		return err
 	}
+	exitNodeBackupResolvers, err := m.os.GetExitNodeForwardResolver()
```
Comment (danderson) [inline net/dns/manager.go:65]: Put this in compileconfig, that's where cfg gets split into rcfg and ocfg.

File: net/dns/osconfig.go [hunk 3]
```diff
@@ -18,20 +19,35 @@ type OSConfigurator interface {
...
 	// after SetDNS has been called to set a Tailscale configuration.
 	// Only works when SupportsSplitDNS=false.
-
+	//
 	// Implementations that don't support getting the base config must
 	// return ErrGetBaseConfigNotSupported.
 	GetBaseConfig() (OSConfig, error)
+
+	// GetExitNodeForwardResolver returns the resolver(s) that should
+	// be used as a fallback for the exit node's DNS-over-HTTP peerapi
```
Comment (danderson) [inline net/dns/osconfig.go:40]: Not a fallback, it's where peerapi requests should go always. Whether global resolvers are configured or not is irrelevant.

File: net/dns/direct.go [hunk 4]
```diff
@@ -183,6 +184,24 @@ func (m *directManager) readResolvFile(path string) (OSConfig, error) {
 	return readResolv(bytes.NewReader(b))
 }
 
+func (m *directManager) GetExitNodeForwardResolver() (ret []dnstype.Resolver, retErr error) {
+	for _, filename := range []string{backupConf, resolvConf} {
+		if oc, err := m.readResolvFile(filename); err == nil {
+			for _, ip := range oc.Nameservers {
+				if ip != netaddr.IPv4(100, 100, 100, 100) {
```
Comment (danderson) [inline net/dns/direct.go:191]: No need to specialcase quad 100. If that's what the exit node uses for DNS, it's perfectly fine for peerapi to loop back into it.

File: net/dns/osconfig.go [hunk 5]
```diff
@@ -90,3 +106,16 @@ func (a OSConfig) Equal(b OSConfig) bool {
...
 // doesn't support reading the underlying configuration out of the OS.
 var ErrGetBaseConfigNotSupported = errors.New("getting OS base config is not supported")
+
+func getExitNodeForwardResolverFromBaseConfig(o OSConfigurator) (ret []dnstype.Resolver, retErr error) {
+	oc, err := o.GetBaseConfig()
+	if err != nil {
+		return nil, err
+	}
+	for _, ip := range oc.Nameservers {
+		if ip != netaddr.IPv4(100, 100, 100, 100) {
```
Comment (danderson) [inline net/dns/osconfig.go:116]: Same, specialcase not needed. peerapi is not in the resolution path for quad-100, therefore sending from peerapi to quad-100 is not a problem, and is desirable.

File: net/dns/resolver/tsdns.go [hunk 6]
```diff
@@ -83,6 +83,14 @@ type Config struct {
 	// LocalDomains is a list of DNS name suffixes that should not be
 	// routed to upstream resolvers.
 	LocalDomains []dnsname.FQDN
+	// ExitNodeBackupResolvers are where the local node when
+	// acting as an exit node and serving a DNS proxy should
+	// forward DNS requests to in the case where there are no
+	// routes found. For example, for Linux systemd-resolved
```
Comment (danderson) [inline net/dns/resolver/tsdns.go:89]: "no routes" ? I don't understand.

**Review by danderson — COMMENTED** [review]
Fairly sure this is overcomplicating things. Just read /etc/resolv.conf and forward to whatever it says, whether or not it's quad-100, whether or not resolved is in charge. That should always produce the correct result on linux, and produces the least incorrect result on other platforms.

**Comment by bradfitz** [comment]
> Fairly sure this is overcomplicating things. 

We don't want to loop back to 100.100.100.100 and get the exit node's MagicDNS names blended in if it's in a different tailnet (later)

File: net/dns/resolver/tsdns.go (hunk 6 above)
Comment (bradfitz) [reply net/dns/resolver/tsdns.go:89]: If the Tailnet DNS config has neither global resolvers or Split DNS routes.

File: net/dns/direct.go (hunk 4 above)
Comment (bradfitz) [reply net/dns/direct.go:191]: We don't want to loop back to 100.100.100.100 and get the exit node's MagicDNS names blended in if it's in a different tailnet.

We can discuss in eng meeting.

**Comment by bradfitz** [comment]
New strategy: https://github.com/tailscale/tailscale/pull/3436

**Pull request closed without merging**